### PR TITLE
Add new Performance API's to October CMS

### PR DESCRIPTION
**Pull request by ayumihamsaki**
This PR adds the new Priority Hints API, Preload API and async to the October CMS AJAX injected code and adds the new Priority Hints API to the loaded assets in the head section.

This is related to the github issues: https://github.com/octobercms/october/issues/4277 and https://github.com/octobercms/october/issues/4030

## Explanation

As performance becomes increasingly important, it's exciting to see browsers implement new features which give developers more control over resource loading. Resource Hints such as `rel=preload` and `rel=preconnect` give developers more control over resource loading and connections to cross-origin servers, respectively. `Client Hints` expose details of a user's device and preferences that developers can use to improve performance in nuanced ways. Continuing in this vein, a new api feature known as `Priority Hints`.

When a browser downloads a resource, the resource is assigned a priority. By default, priorities depend on the type of resource (e.g., script, image, etc.), and the location of the resource reference in the document. For example in Chrome, CSS loaded in typical fashion via the `<link>` element in the `<head>` will be assigned a priority of **highest**, as it blocks rendering. Images in the viewport may be assigned a priority of **high**, whereas images outside the viewport may be assigned a priority of **low**. A `<script>` loaded at the end of the document may receive a priority assignment of **medium** or **low**, but this can be influenced by defer and async.

## Priority Hints API

The API is rather simple and was created by Google. It features an attribute called: `importance` which you can assign either a `high`, `low` or `auto` value to it.

- **high** - The developer considers the resource as being high priority.
- **low** - The developer considers the resource as being low priority.
- **auto** - The developer does not indicate a preference. This also serves as the attribute's invalid value default and missing value default.

In this PR I have given the important assets in the `<head>` section a `importance="high"` and the October AJAX code a `importance="low"`. This is because the important assets are needed to render the web page and the assets for October CMS AJAX code can be loaded after the render of the web page. This increases the load performance of the web page in October CMS.

## Useful Links

Priority Hints API Spec: https://wicg.github.io/priority-hints/

### CODE REVIEW

**Comment by ayumihamsaki** [comment]
Note: There are conflicts here because I have taken the code from this PR https://github.com/octobercms/october/pull/4285 (which hasn't fully been updated yet).

**Comment by ayumihamsaki** [comment]
**Note**: There are conflicts here because I have taken the code from this PR #4285 (which hasn't fully been updated yet).

**Comment by ayumihamsaki** [comment]
**Note**: There are conflicts here because I have taken the code from this PR https://github.com/octobercms/october/pull/4272 (which hasn't fully been updated yet).

**Comment by LukeTowers** [comment]
@ayumihamsaki please resolve conflicts to reflect my latest changes to those files.

**Comment by ayumihamsaki** [comment]
@LukeTowers Thanks, I didn't release you update the code, I will fix this pr accordingly.

I got another quick question, see the following screenshot:

qs

Notice the following filenames:

```
auth.htm
default.htm
form-with-sidebar.htm
```

How come they don't have an underscore in front of them? To be like this:

```
_auth.htm
_default.htm
_form-with-sidebar.htm
```

Then all the filenames will match the same style.

I know this is a minor thing, I'm just use to working with ISO level standard things where everything matches and we have to add `continually` to everything on purpose.

**Comment by ayumihamsaki** [comment]
@LukeTowers Please review my updated commit, I think this resolves the conflicts now ?

**Comment by ayumihamsaki** [comment]
I've had a think on this and wondered if it would also be a good idea to add the `client hints` API as well at the same time.

My personal setup I use the following:

```
accept-ch: Device-Memory, DPR, Width, Viewport-Width
accept-ch-lifetime: 3600
```

These are HTTP Headers by the way, for more details see:

https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/client-hints/
https://developers.google.com/web/updates/2015/09/automating-resource-selection-with-client-hints

Spec: https://httpwg.org/http-extensions/client-hints.html

If it's ok I will add these headers into the two files:

```
modules/backend/layouts/_head.htm
modules/backend/layouts/auth.htm
```

**Comment by w20k** [comment]
@ayumihamsaki I do think `hints` are a good idea to add, but not in this particular use case. 
It's mainly a website specific headers, which should be added manually and not automated.

**Comment by ayumihamsaki** [comment]
@w20k When you say `hints` which one do you mean? There are two different API's here: `Client Hints` and `Priority Hints`. Or maybe you are referring to both. Not sure what you mean also by:

> It's mainly a website specific headers, which should be added manually and not automated.

Are you talking about the `Client Hints` HTTP Header in this case?

**Comment by w20k** [comment]
I was talking about client headers hints, where priority hints is a really  good addition ;)

**Comment by ayumihamsaki** [comment]
Yah that's what I thought. So you want me to remove the `Client Hints` HTTP Header? The reason for adding it is because I have seen some people's plugins that would benefit from the header. Such as a plugin that creates a gallery list of images. Let me know if you want me to delete it out of the code or not? I added it for people's plugins basically.

**Comment by ayumihamsaki** [comment]
@w20k Done if you want to re-view this now, think ready for merge/review process now. Thanks.

**Comment by w20k** [comment]
Done with testing! Might not able to merge it today, but will do it on weekends! Thanks, @ayumihamsaki!

Screenshots:
Screenshot 2019-06-14 at 16 20 51
Screenshot 2019-06-14 at 16 24 54

**Comment by ayumihamsaki** [comment]
@w20k Thanks 🥇

File: modules/cms/twig/FrameworkNode.php [hunk 1]
```diff
@@ -35,25 +35,35 @@ public function compile(TwigCompiler $compiler)
             $compiler
                 ->write("if (\$_minify) {" . PHP_EOL)
                 ->indent()
-                    ->write("echo '<script src=\"'. Request::getBasePath()
-                    .'/modules/system/assets/js/framework.combined-min.js\"></script>'.PHP_EOL;" . PHP_EOL)
+                    ->write("echo '<script async src=\"'. Request::getBasePath()
+                    .'/modules/system/assets/js/framework.combined-min.js\" importance=\"low\"></script>'.PHP_EOL;" . PHP_EOL)
+                    ->write("echo '<link rel=\"preload\" href=\"'. Request::getBasePath()
```
Comment (LukeTowers) [inline modules/cms/twig/FrameworkNode.php]: Why were these added? Many instances of October load the framework assets at the end of the document where <link> would be invalid. Also, how does it have any effect at all appearing immediately after the asset is requested in the first place?

Comment (ayumihamsaki) [reply modules/cms/twig/FrameworkNode.php]: I've tested this and the browser does preload them, have you tested it?

**Comment by ayumihamsaki** [comment]
See Proof:

Untitled

**Comment by LukeTowers** [comment]
Weird, I always thought that <link> had to be in the head. Rereading https://developer.mozilla.org/en-US/docs/Web/HTML/Element/link it appears that's changed.

**Comment by LukeTowers** [comment]
Thanks for the clarification @ayumihamsaki, looks good to me.

**Comment by ayumihamsaki** [comment]
No problem.

I'm going to add some extra info for future reference just in case someone asks this question as well.

With regards to the new performance API the following should be applied (as I done in this PR).

The new performance API now gives us a two level preload rendering, so we get the following:

## Sync Javascript Example

I put the following code in the `<head>` 

```html
<link rel="preload" href="main.js" as="script" importance="high">
<script src="main.js" importance="high"></script>
```

## Async Javascript Example

I put the following code at the bottom of the web page just above the closing body tag

```html
<link rel="preload" href="main.js" as="script" importance="low">
<script async src="main.js" importance="low"></script>
```

Note: This may confuse people when looking at a timing diagram, but the browser will put both the two examples at the top of the waterfall. The Browser will then render preload with the `importance="high"` first and then the browser will render the preload with importance="low"` later.

Hope that helps people to understand.

**Comment by w20k** [comment]
@ayumihamsaki thanks for a `backup` ;) Had a short weekend/vacation)!

**Comment by ayumihamsaki** [comment]
@w20k Nice, hope you went somewhere good!

**Comment by acasar** [comment]
This PR seems to break backwards compatibility.

I am including some custom scripts after the ``{% framework extras %}`` call. It used to work fine (framework scripts were loaded before my custom scripts) but after this PR i'm getting an error:

image

To provide more details - this was the situation before this PR:

image

And this is the situation after the PR, which causes the error above:

image

The fix is of course easy enough, but shouldn't we make sure that existing applications aren't broken? Maybe we should provide a different twig call to include Performance API parameters?

**Comment by acasar** [comment]
Also, due to a mix up in tagging, this PR has already made it to stable release. See more details here: https://github.com/octobercms/october/issues/4465

**Comment by ayumihamsaki** [comment]
Hi, I can't reproduce your error. I did the following:

Test 1:

```
{% framework extras %}
{% scripts %}
        
{# Scripts #}
<script async src="https://www.google-analytics.com/analytics.js" importance="low"></script>
<script async src="{{ 'assets/js/bottom.js'|theme }}" importance="low"></script>
```

Test 2:

```
{# Scripts #}
<script async src="https://www.google-analytics.com/analytics.js" importance="low"></script>
<script async src="{{ 'assets/js/bottom.js'|theme }}" importance="low"></script>

{% framework extras %}
{% scripts %}
```
        
Maybe someone selse can reproduce it? I wait for my more feedback.

**Comment by ayumihamsaki** [comment]
Hi, I can't reproduce your error. I did the following:

Test 1:

```
{% framework extras %}
{% scripts %}
        
{# Scripts #}
<script async src="https://www.google-analytics.com/analytics.js" importance="low"></script>
<script async src="{{ 'assets/js/bottom.js'|theme }}" importance="low"></script>
```

Test 2:

```
{# Scripts #}
<script async src="https://www.google-analytics.com/analytics.js" importance="low"></script>
<script async src="{{ 'assets/js/bottom.js'|theme }}" importance="low"></script>

{% framework extras %}
{% scripts %}
```
        
Maybe someone else can reproduce it? I wait for my more feedback.

This is probably a stupid question, but do you have a conflict in your javascript instead. Because the preload function is loading that code in a different order now.

**Comment by acasar** [comment]
I think the key is to try without ``async`` and ``importance="low"`` attributes on your custom scripts.

When using the attributes it of course works, but most existing October applications don't use them, that's when the issues happen.

**Comment by ayumihamsaki** [comment]
Your code order from the screenshots you have given is like this:

Before your code is loading like this:

```
combine.js
framework.combined-min.js
framework.extras.min.css
..
```

Now after preloading the order is like this:

```
framework.combined-min.js
framework.extras-min.css
combine.js
..
```

**Comment by ayumihamsaki** [comment]
Your code order from the screenshots you have given is like this:

Before your code is loading like this:

```
combine.js
framework.combined-min.js
framework.extras.min.css
..
```

Now after preloading the order is like this:

```
framework.combined-min.js
framework.extras-min.css
combine.js
..
```

Try preloading your `combine.js` file.

stretch

Let me know if this works or not?

**Comment by acasar** [comment]
@ayumihamsaki the issue is in frontend.js, not combine.js , but I already said that it's an easy thing to fix and it's not the reason I posted my comment.

The main issue is that it breaks backwards compatibility. Usually just updating October to the latest release shouldn't require any code changes, but after this goes live, most applications that include custom javascript without any preloading attributes will need to update their code.

So to sum up -  I understand this change and I'm not against it, but we should be able to make this opt-in and not applied by default since it clearly breaks existing applications.

**Comment by ayumihamsaki** [comment]
@acasar understand, sorry we should of added a comment in the update release about this breaking change.

**Comment by acasar** [comment]
In my opinion this is to big of a change to simply solve it with a release note. At my company alone we'd need to update about 90 websites, but I believe thousands of websites will be affected overall as well as most of the themes available in the October marketplace.

My suggestion is that we leave ``{% framework extras %}`` intact so that it works as it always did and introduce a new tag (e.g. ``{% framework optimized-extras %}``) to include performance related attributes.

**Comment by ayumihamsaki** [comment]
I know Luke won't go for a new tag called: `{% framework optimized-extras %}` we discussed this in the past.

But don't get me wrong, every pr I submit I always want to cater for backwards compatibility, will wait for feedback from one of the admins.

**Comment by Samuell1** [comment]
Maybe better will be parameter, something like: `{% framework extras preload %}`?

@ayumihamsaki Btw why scripts in backend are loaded in head not at end of body to prevent it to block render?

**Comment by LukeTowers** [comment]
@ayumihamsaki If you could make a PR to revert the change to `{% framework extras %}` to require an extra `async` (`{% framework extras async %}`) also do it for `{% framework %}` and `{% jquery %}` so we can roll it into 458 to make the builds stable that would be appreciated. @acasar is correct, this shouldn't have made it in by default, I didn't realize it would break sites in that way.

**Comment by LukeTowers** [comment]
@bennothommo @daftspunk This is going to require fixing before we can make 457 stable.

**Comment by ayumihamsaki** [comment]
@LukeTowers 

I think,

{% framework extras preload %}

is better than

{% framework extras async %}

The async issue is because of the browser version, using a combination of `async` and `importance="low"` and so it's behind a flag at the moment for older browsers - sorry I forgot to mention that part.

For newer browsers which don't need to turn on the flag the issue would be a `preload` order. So I agree with @Samuell1 with the naming.

**Comment by LukeTowers** [comment]
@ayumihamsaki I don't, preload implies that you're going to make it load before the rest of the scripts which is exactly the opposite of the problem here. The problem is that the scripts no longer load in the same order that was previously expected because they are being loaded asynchronously. I get that the `preload` attribute may be the more important for performance but the `aysnc` attribute is what causes the issue so I think that's what we should emphasize.

**Comment by ayumihamsaki** [comment]
@LukeTowers you're not correct, you are looking at the code from a normal async view and not using the importance API. It will be like this:

```html
<script async src=/modules/system/assets/js/framework.combined-min.js importance="low"></script>
<link preload href=/modules/system/assets/js/framework.combined-min.js as="script" importance="low">
```

### Importance API turned on (flag on) - lastest Google Canary:

The files are loaded `sync` not `async` due to combination of the `importance` API and `preload`.

### Importance API turned off (flag off) - stable Google Chrome:

The files are loaded `async` due to the importance API not turned on.

**Comment by LukeTowers** [comment]
@ayumihamsaki fair enough. We can use `preload` then but in the docs we should make note of the fact that `preload` will load assets `async` in older browsers so that people are aware of the impacts of using it.

**Comment by Samuell1** [comment]
@ayumihamsaki Btw why this change was made to "front end" twig parser (modules/cms/twig/FrameworkNode.php) ? Isnt for now best to remove it and developer should choose how he wants to import framework?

**Comment by ayumihamsaki** [comment]
Yah good idea. Sorry this is totally my fault for using Google Canary and not thinking about the experiments flag turned on.

**Comment by ayumihamsaki** [comment]
@LukeTowers sorry to keep pinning you, I just had a chat with Addy Osmani one of the importance API authors, he was saying that right now different browsers are doing different things with regards to `async` and `prioritized`.

He suggested the best idea woul;d be like this:

```
{% framework extras optimized %}
```

Then explain in the doc's about the preloading and async/sync actions by the browsers - modern and backwards compatibility.

**Comment by ayumihamsaki** [comment]
I have created a PR fix here:

https://github.com/octobercms/october/pull/4463/commits/00f688e6b7ce95eea26c9875a888a57721761a2d

We now have the following three options:

```
{% framework %}
{% framework extras %}
{% framework extras_optimized %}
```

Very sorry to @acasar for the issue!

**Comment by ayumihamsaki** [comment]
I have created a PR fix here: https://github.com/octobercms/october/pull/4463/commits/00f688e6b7ce95eea26c9875a888a57721761a2d

We now have the following three options:

```
{% framework %}
{% framework extras %}
{% framework extras_optimized %}
```

Very sorry to @acasar for the issue!

**Comment by acasar** [comment]
@ayumihamsaki no problem at all :)

Thanks to everyone for looking into this so quickly 🥇

**Comment by LukeTowers** [comment]
@ayumihamsaki why is it `_optimized` instead of a separate option entirely? I was thinking a separate option so that `{% framework optimized %}` and `{% jquery optimized %}` and `{% framework extras optimized %}` would all be options.

**Comment by ayumihamsaki** [comment]
@LukeTowers it gave me an error when testing.

```
{% framework extras optimized %} = Error
{% framework extras-optimized %} = Error
{% framework extras_optimized %} = Passed
```

I can create the following tomorrow:

```
{% framework %}
{% framework_optimized %}
{% framework extras %}
{% framework extras_optimized %}
```

p.s. What do you mean by `{% jquery optimized %}`, are you thinking so just creating a minified version ? If so why not just minify all the files by default.

Let me know I will update the pr later.

**Comment by bennothommo** [comment]
@ayumihamsaki I personally am not a fan of the underscores - what sort of error are you getting with `extras optimized`?

**Comment by w20k** [comment]
Thanks for feedback @acasar! 
My bad, need to include more front-end tests for my test routine!

**Comment by Samuell1** [comment]
@bennothommo Its returning some weird twig error `Unexpected token "name" of value "optimized" ("end of statement block" expected).`

**Comment by ayumihamsaki** [comment]
Sorry late reply, had to get some sleep. yah exactly what Samuell1 said is what I was seeing. That's why I went for the underscore as computers read `extras_optimized` and `extrasoptimized` the same.

**Comment by ayumihamsaki** [comment]
Adding extra security, I plan to add Subresource Integrity to the loaded jQuery files via the tags.

**Comment by ayumihamsaki** [comment]
Adding extra security, I plan to add Subresource Integrity to the loaded jQuery files via the tags.

There is a small catch and I will do the following to not mess things up!

I will add the current version of `jQuery v3.3.1` and `jQuery Migrate v3.0.1`, to this pr.

As Luke has put the following pr on hold: https://github.com/octobercms/october/pull/4461

I will update the `FrameworkNode.php` file with a newer sri to match those new versions of `jquery` and `jquery migrate`, then everything will go smoothly and people won't even notice.

**Comment by ayumihamsaki** [comment]
@bennothommo 

Sorry late reply, had to get some sleep, yah exactly what Samuell1 said is what I was seeing. That's why I went for the underscore as computers read `extras_optimized` and `extrasoptimized` the same.

If you want to fix the parser side to allow mutli-parms. I will update the pr to this:

```
{% framework %} 
{% framework optimized %}
{% framework extras %}
{% framework extras optimized %}
{% framework jquery %} = loads jquery sync only (not framework)
{% framework jquery optimized %} = loads jquery sync, high priorty and preload (not framework)
```

Updated commit found here: https://github.com/octobercms/october/pull/4463/commits/7aefdcf5b741c772f591cfa9842ec71a59b76ab8

**Comment by ayumihamsaki** [comment]
Notes for people reading this in the future.

The twig tags load jQuery `sync` not `(async)`. If you want to load jQuery using `async` you can do so by manually adding the jquery into your web page without using the tags.

For further info how to setup jQuery `async` you can read this article: https://idiallo.com/javascript/async-jquery

**Comment by bennothommo** [comment]
@ayumihamsaki No problem :). If a fix to the argument parser can be done purely within October code, then go for it. If it will require an update to the Twig framework directly, then don't worry about it - any changes to the Twig code may be lost with an update, and I would imagine they have their reasons for not allowing it.

**Comment by Samuell1** [comment]
This twig parser file specific for framework tag needs to be changed somehow to allow passing $name and attributes too: https://github.com/octobercms/october/blob/develop/modules/cms/twig/FrameworkTokenParser.php

**Comment by ayumihamsaki** [comment]
> If it will require an update to the Twig framework directly, then don't worry about it.

Going to set them to this:

```
{% framework %} 
{% framework optimized %}
{% framework extras %}
{% framework extras_optimized %}
{% framework jquery %} = loads jquery sync only (not framework)
{% framework jquery_optimized %} = loads jquery sync, high priorty and preload (not framework)
```

As per commit: https://github.com/octobercms/october/blob/7fd15432cf9af5e756de213cc8e453ef7053975c/modules/cms/twig/FrameworkNode.php

**Comment by LukeTowers** [comment]
This has become too complicated and I'm not satisfied with the experience of using these "optimized" tags. @ayumihamsaki please create a PR to completely revert your changes to the framework, framework extras, and jquery tags. If we still want to create a way for developers to easily opt in to the optimized versions then that can be a separate issue / PR, but for now I just want the backwards incompatible issue solved so we can release 458. Any developers that absolutely need the optimized version can always include those files however they want to, these tags are just shortcuts.

**Comment by ayumihamsaki** [comment]
@LukeTowers no problem, I will revert the changes now. If I hard code the files into my websites will there be any danger of filename changes or anything in the future?

**Comment by LukeTowers** [comment]
@ayumihamsaki I doubt it. If there are they'd be noted in the release notes.

**Comment by ayumihamsaki** [comment]
@LukeTowers 

Done:

Revert file is here: https://github.com/octobercms/october/blob/4a03be9f90456189ce619e762c3583e32f131f8b/modules/cms/twig/FrameworkNode.php

Gives the following:

```
{% framework %} 
{% framework extras %}
{% framework jquery %}
```

Updated doc's here: https://github.com/octobercms/docs/blob/ad0a47a9cf160c38c0d2ef331564bf1443dc07f4/ajax-introduction.md#framework-options

p.s. Going to try my best to finish off the large PR tomorrow, as I know you want to push this commit quickly.

**Comment by LukeTowers** [comment]
@ayumihamsaki can you split those two changes into their own PRs? I only want to merge those changes right now, the big changes will take a lot more reviewing before they're ready to merge.

**Comment by Samuell1** [comment]
Btw if framework tag changes will be made later it needs to reflect on combiner aliases too. https://octobercms.com/docs/markup/filter-theme#combiner-aliases 

And from my experience i mostly recommend importing scripts like that because then you can combine many requests to few files.

**Comment by ayumihamsaki** [comment]
Nice tip!

Want to change the following doc's line:

```
@jquery | Reference to the jQuery library (v3.4.0) used in the back-end. (JavaScript)
```

into

```
@jquery | Reference to the jQuery library (highest available version) used in the back-end. (JavaScript)
```

That way I don't need update this doc's page every time a jquery update takes place.

**Comment by LukeTowers** [comment]
@Samuell1 actually the aliases won't need to be touched, the "optimized" tag changes how the script / style / link elements are generated, not the actual content itself. If someone's using the combiner aliases then they should be doing those optimizations themselves. @ayumihamsaki I'd prefer we kept the version information there, but you could maybe add a note to the files in the october repo to make sure that we always update that information when updating the jQuery version.

**Comment by ayumihamsaki** [comment]
> but you could maybe add a note to the files in the october repo to make sure that we always update that information when updating the jQuery version.

Yah going to make a note of that for future ref.

**Comment by LukeTowers** [comment]
@ayumihamsaki @Samuell1 @acasar The fix has been implemented in https://github.com/octobercms/october/commit/99777f5977ef50d1b8757bf0fc1e539e5e8a5afb. We can discuss adding the optimized versions in a separate issue dedicated issue just for that, but seeing as it's just a shortcut and devs can always write that version manually into their themes themselves I don't really see a huge need for it right now.

**Comment by ayumihamsaki** [comment]
I'm NOT up for adding the optimized versions! I just created a partial file with the optimized versions and  added that twig into my layout files instead.

**Comment by LukeTowers** [comment]
@ayumihamsaki fair enough, that seems like the easiest solution for people that want to do it that way 😂 Thanks for your work on this so far though!
